Add support for MSP-EXP432P401R-LAUNCHXL

Mani-Sadhasivam opened the pull request:
This PR adds support for Texas Instruments MSP-EXP432P401R-LAUNCHXL development board. PR has been divided into following patches:

1. SoC Patch:
* Added support for MSP432P401R SoC
* Added dts support
* Added MSP432P4XX SDK support

2. Board Patch;
* Added support for MSP-EXP432P401R-LAUNCHXL board
* Added documentation for the board

3. UART Patch:
* Added support for on-board UART_0

More support for other on-board peripherals will be added soon

Mani-Sadhasivam committed:
arch: arm: soc: ti_simplelink: Add MSP432P401R SoC

This patch adds SoC support for TI SimpleLink family MSP432P401R.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
boards: arm: Add support for TI MSP-EXP432P401R-LAUNCHXL

This patch adds support for TI Simplelink MSP-EXP432P401R-LAUNCHXL
development board based on Cortex M4 family

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Mani-Sadhasivam committed:
drivers: serial: Add MSP-EXP432P401R UART

This patch adds support for on board UART_0 on MSP-EXP432P401R-LAUNCHXL.
Driver makes use of driverlib available in ROM by default, thus saving
code space.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
+The SimpleLink MSP‐EXP432P401R LaunchPad development kit is an easy-to-use evaluation
+module for the SimpleLink MSP432P401R microcontroller. It contains everything needed to start
+developing on the SimpleLink MSP432 low-power + performance ARM |reg| 32-bit Cortex |reg|-M4F
+microcontroller (MCU). The MSP432P401R device supports low-power applications requiring increased CPU
```
"increased" implies comparison with something else, but what are you comparing to?  And I'm not sure what "32-bit performance" means.

dbkinder commented on MAINTAINERS:
```diff
@@ -119,6 +119,11 @@ S:      Supported
 F:      boards/arm/cc3200_launchxl/
 F:      boards/arm/cc3220sf_launchxl/
 
+BOARDS/ARM - TI MSP-EXP432P401R LAUNCHXL
```
@nashif are we shifting over to using CODEOWNERS and not MAINTAINERS?

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
++-----------+------------+-----------------------+
+| Interface | Controller | Driver/Component      |
++===========+============+=======================+
+| NVIC      | on-chip    | nested vectored		 |
```
table columns need to line up (with spaces)

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
+
+	$ make -C samples/hello_world BOARD=msp_exp432p401r_launchxl
+
+This will produce zephyr.elf which could be flashed onto MSP-EXP432P401R LaunchXL
```
How about, 

     The resulting ``zephyr.elf``` binary can be flashed ...

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
+For Linux:
+----------
+
+`UniFlash`_ command line utility could be used to program the flash memory. Only
```
s/could be/is/

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
+`UniFlash`_ command line utility could be used to program the flash memory. Only
+elf loading is supported as of now.
+
+The following command has been known to work:
```
how about:

     The following command line will flash the ``zephyr.elf`` binary to the LaunchXL board:

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,112 @@
+
+References
+**********
+
```
Normally the link name such as

     .. _MSP-EXP432P401R LaunchXL User's Guide:
        http://www.ti.com/lit/ug/slau597c/slau597c.pdf

won't show up in the generated output.  If you'd like them to show up here in this "References" section, you can add a directive to have them show up right here (add a blank line before and after this directive):

     .. target-notes::

dbkinder reviewed (changes requested):
suggested doc edits

Mani-Sadhasivam commented:
Thanks a lot @dbkinder for your review. I'll incorporate your suggestions.

galak reviewed (changes requested):
Please split out the HAL import into its own commit, than a commit that adds the Kconfig/building support for the HAL after that.

galak commented on MAINTAINERS:
```diff
@@ -211,6 +216,16 @@ M:      Gil Pitney <gil.pitney@linaro.org>
+S:      Supported
+F:      ext/hal/ti/msp432p4xxsdk/
+
+MSP432P4XX SOC - TI SIMPLELINK
```
should be the SOC Patch

Mani-Sadhasivam replied:
Done

Mani-Sadhasivam replied:
Done

galak commented on MAINTAINERS:
```diff
@@ -211,6 +216,16 @@ M:      Gil Pitney <gil.pitney@linaro.org>
 S:      Supported
 F:      arch/arm/soc/ti_simplelink/cc32xx
 
+MSP432P4XX SDK
```
Should be in the HAL/SDK import patch (the one that adds Makefile/Kconfig)

Mani-Sadhasivam replied:
Done

Mani-Sadhasivam replied:
Done

nashif reviewed (dismissed):
In the commit introducing the external components into ext/ please add some details about the origin of the code in the commit message, see the contribution guidelines for the required information:

https://github.com/zephyrproject-rtos/zephyr/wiki/Contribution-Guide#change-requirements

Mani-Sadhasivam commented:
@nashif Commit message has been added as per the guidelines

Mani-Sadhasivam commented:
@galak @nashif I have addressed all review comments so far. Is there anything lacking from my end?

galak commented on ext/hal/ti/msp432p4xxsdk/Kbuild:
```diff
@@ -0,0 +1,14 @@
+obj-$(CONFIG_MSP432P4XXSDK_BUILTIN) += driverlib/interrupt.o
+obj-$(CONFIG_MSP432P4XXSDK_BUILTIN) += driverlib/system_msp432p401r.o
+obj-$(CONFIG_MSP432P4XXSDK_BUILTIN) += driverlib/cpu.o
+obj-$(CONFIG_MSP432P4XXSDK_BUILTIN) += driverlib/i2c.o
```
the i2c, uart, spi should be based on the driver being enabled.

galak reviewed (changes requested):
Any reason we can't use SDK version 1.40.01.00 that came out recently?

galak commented on drivers/serial/uart_msp432p4xx.c:
```diff
@@ -0,0 +1,295 @@
+	MAP_GPIO_setAsPeripheralModuleFunctionInputPin(GPIO_PORT_P1,
+		(GPIO_PIN2 | GPIO_PIN3), GPIO_PRIMARY_MODULE_FUNCTION);
+
+	/* 115200 Buad rate settings calculated for 48MHz */
```
We should be getting baud rate from the device tree.

This patch should have a YAML associated with it.

galak commented on drivers/serial/uart_msp432p4xx.c:
```diff
@@ -0,0 +1,295 @@
+#endif
+
+static const struct uart_device_config uart_msp432p4xx_dev_cfg_0 = {
+	.base = (void *)EUSCI_A0_BASE,
```
base should come from Device tree.

dbkinder reviewed (dismissed):
+1 for docs

GAnthony reviewed (dismissed):
Looks generally good.  I see UART is different than CC32xx, so cannot simply reuse that driver.

I added a few comments on imported SDK directory structure, and request instruction on how to debug this board.

Mani-Sadhasivam commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,111 @@
+The resulting ``zephyr.elf`` binary can be flashed onto MSP-EXP432P401R LaunchXL
+using the command line utility mentioned below.
+
+Flashing
```
@GAnthony Thanks for your review. 

Debugging this board through command line interface is not possible AFAIK. Only way I can debug is by using CCS IDE .

Do you want me to add instructions for CCS debugging?

GAnthony replied:
For gdb debugging, I found this: http://software-dl.ti.com/msp430/msp430_public_sw/mcu/msp430/MSP432GCC/latest/index_FDS.html
and this:
http://www.ti.com/lit/ug/slau688b/slau688b.pdf

So maybe can say that debugging is primarily supported via CCS, and link to the appropriate guide/section; but can mention there also exists a gdb/gcc method, which is also available.

Mani-Sadhasivam replied:
I tried to debug by following those guides. But there is no xds2xxu_msp432_swd.dat file exist with the package. Anyway, I can add guides for both CCS and gdb as you said and will report this in TI forum also.

Thanks for pointing out!

Mani-Sadhasivam commented on ext/hal/ti/msp432p4xxsdk/README:
```diff
@@ -0,0 +1,33 @@
+
+	http://www.ti.com/tool/simplelink-msp432-sdk
+
+Files in driverlib/ and inc/ are copied from an SDK installation
```
Makes sense! I guess this directory structure can be adopted once PR has been accepted as that will be more easy.

GAnthony replied:
Sure.  Another PR  #789 should be coming in soon, which will test our ability to have these SDKs splice together as TI has recently restructured :-)

Mani-Sadhasivam replied:
That would be nice :-)

Mani-Sadhasivam commented:
@dbkinder I've added debugging instructions based on review from @GAnthony. Please take a look at that

Mani-Sadhasivam commented:
@galak I've incorporated your suggestions as below:

1. Changed SDK to latest version 1.40.01.00
2. Modified baud rate and base address setting for UART based in config parameters
3. Addded UART yaml to UART patch (68a3e9870c10efaf2b309b0bdf0bdf9ad34bb158)

galak reviewed (changes requested):
As @GAnthony we should put the SDK in ext/hal/ti/simplelink/source/ti/devices/msp432p4xx

Also, the SDK import should be the first patch in the series.

Mani-Sadhasivam commented:
@galak @GAnthony Sure I'll update SDK. 

However I'd like to get the following concerns addressed for merging both SDK's together under ext/hal/ti/simplelink/

1. cc3220 build doesn't allow using the sources inside driverlib/, it only has the provision to use ROM 
    drivers but msp432p4xx SDK allows using both. So, if I merge msp432p4xx SDK build support then
    we'll have two devices of same family offering different functionalities. Is this fine for you?

2. Current README under ext/hal/ti/simplelink/ is not generic. So, for adding MSP432 SDK can I create 
    two sections to isolate both SDK's?. Later we can see how to make this README generic for 
    simplelink family.

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,128 @@
+| Interface | Controller | Driver/Component      |
++===========+============+=======================+
+| NVIC      | on-chip    | nested vectored       |
+|                        | interrupt controller  |
```
need the vertical bar `|` column separator on this line (aligned with the one on the previous line)

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,128 @@
+----------
+
+`UniFlash`_ command line utility is used to program the flash memory. Only
+elf loading is supported as of now.
```
s/supported as of now/currently supported/

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,128 @@
+
+.. code-block:: console
+
+   % ./dslite.sh --config=MSP432P401R.ccxml zephyr.elf
```
we're using a dollar sign `$` for the linux prompt character

dbkinder commented on boards/arm/msp_exp432p401r_launchxl/doc/msp_exp432p401r_launchxl.rst:
```diff
@@ -0,0 +1,128 @@
+
+   % ./dslite.sh --config=MSP432P401R.ccxml zephyr.elf
+
+.. note:: ccxml file is included in the MSP432 SDK
```
How about:

     The ccxml configuration file is included in the MSP432 SDK.  See 
     ``ext/hal/ti/msp432p4xxsdk/README`` for more information.

dbkinder reviewed (dismissed):
suggested edits

Mani-Sadhasivam commented:
@GAnthony @galak  Any suggestion for my query?

Merging both MSP432 and CC3220 SDK config together looks messy unless we have some common SDK Kconfig option like CONFIG_HAS_SIMPLELINKSDK. I guess we can sort this out after merging this PR (having SDK in ext/hal/ti/mspp4xxsdk)!

GAnthony commented:
@Mani-Sadhasivam, regarding the queries:

1) For CC3220SDK, I decided to reduce the number of build options, and only provide the ROM apis,  which reduced the number of driverlib files imported into the Zephyr tree.   I also found I wasn't ever needing to build the driverlib files.  I can recommend you also do the same for MSP432 SDK, as a simplification.

2) README can have a section for CC32XX and MSP432, and Kbuild can also be block sectioned using ifeq syntax:
ifeq ($(CONFIG_HAS_CC3220SDK),y)
<build CC3220SDK stuff only>
endif

Other than the README, Makefile, and Kbuild, are there other conflicts between CC32XX SDK and MSP432XX SDK?

Mani-Sadhasivam committed:
ext: hal: Add MSP432P4XX SDK Kconfig/Build support

This patch adds Kconfig/build support to MSP432P4XX
SDK. MSP432P4XX microcontrollers have driverlib flashed
onto the ROM, which will be used by default.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Mani-Sadhasivam commented:
@galak @GAnthony 

Following is the changeset:

1. Moved SDK to ext/hal/ti/simplelink/sources/ti/devices
2. Removed SDK built-in support as suggested by @GAnthony 
3. Fixed doc based on suggestions from @dbkinder

GAnthony commented on ext/hal/ti/simplelink/Kbuild:
```diff
@@ -35,3 +36,10 @@ obj-$(CONFIG_SIMPLELINK_HOST_DRIVER) += kernel/zephyr/dpl/MutexP_zephyr.o
 obj-$(CONFIG_SIMPLELINK_HOST_DRIVER) += kernel/zephyr/dpl/HwiP_zephyr.o
+endif
+
+ifeq ($(CONFIG_HAS_MSP432P4XXSDK),y)
```
Perhaps, since now there is an ifeq guard, the ccflags- and obj-$(CONFIG_HAS...) can be replaced by ccflags-y and obj-y everywhere?

Mani-Sadhasivam replied:
Done

GAnthony commented on arch/arm/soc/ti_simplelink/msp432p4xx/Kconfig.defconfig.msp432p401r:
```diff
@@ -0,0 +1,29 @@
+	int
+	default 64
+
+config MSP432P4XXSDK_LIBRARY
```
Is MSP432P4XXSDK_LIBRARY config variable used anywhere?

Mani-Sadhasivam replied:
Oops! It was added based on CC3200 config. 

Not used anywhere, removed.

Mani-Sadhasivam commented:
@GAnthony Looks like you have reverted the Simplelink host driver commits. I already rebased my change on top of that, so experiencing the conflict. Should I update my PR now or do you have any plans to merge those commits again soon?

GAnthony commented:
Yes, I recommend updating the PR.  

It may be some time before we figure out how to deal with that host driver source.

Sorry for the extra work caused by the reversion.

Mani-Sadhasivam commented:
Thanks I'll update my branch. Btw can you take a look at the changes I've done based on your review?

GAnthony commented:
Those changes look good to me.  Thnx.

Mani-Sadhasivam commented:
@galak I've addressed all of the review comments so far.

galak committed:
libc: minimal: Add typedefs for fast types

Some of the HALs in ext require fast types to be defined.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
ext: hal: ti: Add MSP432P4XX SDK

MSP432P4XX SDK is a CMSIS based SDK offered by Texas Instruments
for supporting their Simplelink MSP432P4XX microcontrollers. The
version included in Zephyr is V1.50.00.12 which is a stripped
down version containing only HAL and SoC header support.

Origin: SIMPLELINK-MSP432-SDK
License: BSD 3-Clause
URL: http://www.ti.com/tool/simplelink-msp432-sdk
Version: V1.50.00.12
Purpose: Provides HAL support to MSP432P4XX SoC
Maintained-by: External

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

galak committed:
ext: hal: ti: msp432p4xx: Tweak SDK to build with Zephyr

There are some minor conflicts or porting changes needed to get the
msp432p4xx SDK to work with Zephyr:
1. Remove definition of BIT, as it conflicts with Zephyr defined one
2. Set __SYSTEM_CLOCK define based on Zephyr Kconfig define
3. Fix a build warning that gets treated as an error.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
@Mani-Sadhasivam updated to latest SDK and cleaned up a bunch of stuff.  Can you take a look and make sure this still runs on hardware.  Thanks

Mani-Sadhasivam commented:
@galak I am not able to see UART output now. Debugging the issue now, will let you know. Thanks for your help :-)

galak commented:
Any luck?

Mani-Sadhasivam commented:
@galak Aah not yet. Didn't have much time to look into it. Will dig through today

Mani-Sadhasivam commented:
@galak So, the issue is with fast integer types. My original revision depended newlibc for fast integer types. But you have removed that dependency and defined those in minimal libc header. 

During debugging I found that all fast integer types (fast8_t, fast16_t...) should be treated as unsigned int for correct working, that's what happening if we add newlib dependency.

So, should we just include CONFIG_NEWLIB_LIBC=y in board defconfig or you have any other suggestions?

galak commented:
We should just fix lib/libc/minimal/include/stdint.h to use 'int' & 'unsigned int' for {u}int_fast{8,16}_t

Mani-Sadhasivam commented:
@galak Changeset:

1. Fixed 'lib/libc/minimal/include/stdint.h' by making all fast types as integer and added a comment
2. Few typo fixes

With the above mentioned changes, I confirm that the code is working fine on HW

galak commented on lib/libc/minimal/include/stdint.h:
```diff
@@ -42,11 +42,23 @@ typedef signed short        int16_t;
+typedef signed int          int_fast8_t;
+typedef signed int          int_fast16_t;
+typedef signed int          int_fast32_t;
+typedef signed int          int_fast64_t;
```
int_fast64_t should be signed long long

galak commented on lib/libc/minimal/include/stdint.h:
```diff
@@ -42,11 +42,23 @@ typedef signed short        int16_t;
+typedef unsigned int        uint_fast8_t;
+typedef unsigned int        uint_fast16_t;
+typedef unsigned int        uint_fast32_t;
+typedef unsigned int        uint_fast64_t;
```
uint_fast64_t should be 'unsigned long long'

galak reviewed (changes requested):
the fast types for 64-bit data should be long long

Mani-Sadhasivam commented:
@galak Changed the 64bit types to long long (was just following newlib before). 

Not sure what do you mean by the issue with commit message as all the checks have passed!
